File: client/ios/Data/Attachment/AttachmentWriter.swift
import Foundation

public protocol AttachmentWriter {
    func attachAttachment(request: AttachmentLinkRequest) async throws -> AttachmentLinkPayload
    func fileHandoff(attachmentId: String) async throws -> AttachmentFileHandoffPayload
    func uploadHandoff(request: AttachmentUploadHandoffRequest) async throws -> AttachmentUploadHandoffPayload
    func uploadAttachment(request: AttachmentUploadHandoffRequest, fileName: String, mimeType: String, data: Data) async throws -> AttachmentItemPayload
}

File: client/android/Data/Attachment/AttachmentWriter.kt
package app.mobiling.client.data.attachment

import app.mobiling.client.contract.attachment.AttachmentFileHandoffPayload
import app.mobiling.client.contract.attachment.AttachmentItemPayload
import app.mobiling.client.contract.attachment.AttachmentLinkPayload
import app.mobiling.client.contract.attachment.AttachmentLinkRequest
import app.mobiling.client.contract.attachment.AttachmentUploadHandoffPayload
import app.mobiling.client.contract.attachment.AttachmentUploadHandoffRequest

/**
 * Marketing America Corp. Oleksandr Tishchenko
 */
interface AttachmentWriter {
    suspend fun attachAttachment(request: AttachmentLinkRequest): AttachmentLinkPayload
    suspend fun fileHandoff(attachmentId: String): AttachmentFileHandoffPayload
    suspend fun uploadHandoff(request: AttachmentUploadHandoffRequest): AttachmentUploadHandoffPayload
    suspend fun uploadAttachment(
        request: AttachmentUploadHandoffRequest,
        fileName: String,
        mimeType: String,
        bytes: ByteArray,
    ): AttachmentItemPayload
}
